### PR TITLE
strategy/record : implement #302 record function

### PR DESCRIPTION
**Pull request by tony1223**
implment for #302

### CODE REVIEW

File: migrations/mysql/20211211034819_add_asset_balance_history_detail.sql [hunk 1]
```diff
@@ -0,0 +1,27 @@
+-- +up
+-- +begin
+create table asset_balance_history_detail
+(
+    gid        bigint unsigned auto_increment
+        primary key,
+    exchange   varchar(30)                                not null,
+    subAccount varchar(30)                                not null,
```
Comment (c9s) [inline migrations/mysql/20211211034819_add_asset_balance_history_detail.sql:8]: we use snake case for the column names, e.g., `subaccount`, `in_btc`, `in_usd`

File: migrations/sqlite3/20211211034818_add_asset_balance_history_detail.sql [hunk 2]
```diff
@@ -0,0 +1,26 @@
+-- +up
+-- +begin
+CREATE TABLE `asset_balance_history_detail`
+(
+    gid    bigint unsigned auto_increment primary key,
+    `exchange`          VARCHAR             NOT NULL DEFAULT '',
+    `subAccount`          VARCHAR             NOT NULL DEFAULT '',
+    time   DATETIME(3)     NOT NULL,
+    currency     VARCHAR                                not null,
+    balanceInUSD DECIMAL(32, 8) default 0.00000000 not null,
```
Comment (c9s) [inline migrations/sqlite3/20211211034818_add_asset_balance_history_detail.sql:10]: upper case for the keywords? `DEFAULT` and `NOT NULL`

File: migrations/mysql/20211211034819_add_asset_balance_history_detail.sql [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    time       datetime(3)                                not null,
```
Comment (c9s) [inline migrations/mysql/20211211034819_add_asset_balance_history_detail.sql:9]: set default to NOW()?

File: migrations/sqlite3/20211211034818_add_asset_balance_history_detail.sql [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    balanceInBTC decimal(32, 8) default 0.00000000 not null,
+    balance    decimal(32, 8) default 0.00000000 not null,
+    available  decimal(32, 8) default 0.00000000 not null,
+    locked     decimal(32, 8) default 0.00000000 not null
```
Comment (c9s) [inline migrations/sqlite3/20211211034818_add_asset_balance_history_detail.sql:14]: decimal(32, 8) will be converted into DECIMAL in sqlite3

File: pkg/exchange/ftx/exchange.go [hunk 5]
```diff
@@ -96,6 +106,18 @@ func (e *Exchange) NewStream() types.Stream {
...
+		return nil, err
+	}
+	marketMap := types.MarketMap{}
+	for k, v := range markets {
+		marketMap[k] = v.Market
+	}
+	return marketMap, nil
+}
+
+func (e *Exchange) _QueryMarkets(ctx context.Context) (MarketMap, error) {
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:120]: go uses lower case method name for private func, you can use something like `queryMarkets`

File: pkg/exchange/ftx/exchange.go [hunk 6]
```diff
@@ -461,11 +488,68 @@ func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) erro
...
+		return nil, err
+	}
+	ticker := ticketMap[symbol]
+	return &ticker, err
 }
 
 func (e *Exchange) QueryTickers(ctx context.Context, symbol ...string) (map[string]types.Ticker, error) {
-	panic("implement me")
+
+	logrus.Debugf("QueryTickers %s ", symbol)
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:502]: remove debug logs

File: pkg/exchange/ftx/exchange.go [hunk 7]
```diff
@@ -461,11 +488,68 @@ func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) erro
...
+		//ctx context.Context, market string, interval types.Interval, limit int64, start, end time.Time
+		prices, err := rest.HistoricalPrices(ctx, v.Market.LocalSymbol, types.Interval1h, 1, time.Now().Add(time.Duration(-1)*time.Hour), time.Now())
+		if err != nil || !prices.Success || len(prices.Result) == 0 {
+			continue
+			//panic(fmt.Sprintf("ftx QueryTickers fail : %s ", k))
+		}
+
+		lastCandle := prices.Result[0]
+		time2 := lastCandle.StartTime.Format(time.UnixDate)
+		_ = time2
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:539]: time2 is not used?

File: pkg/service/account.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	//"fmt"
+	//"strconv"
+	//"strings"
+	//"time"
+	//
+	//"github.com/jmoiron/sqlx"
+	//"github.com/pkg/errors"
+	//log "github.com/sirupsen/logrus"
+	//
+	//"github.com/c9s/bbgo/pkg/exchange/batch"
```
Comment (c9s) [inline pkg/service/account.go:16]: remove unused imports?

File: migrations/mysql/20211211034819_add_asset_balance_history_detail.sql [hunk 9]
```diff
@@ -0,0 +1,27 @@
+-- +up
+-- +begin
+create table asset_balance_history_detail
```
Comment (c9s) [inline migrations/mysql/20211211034819_add_asset_balance_history_detail.sql:3]: `nav_history_details` ? net asset value details

File: migrations/mysql/20211211034819_add_asset_balance_history_detail.sql [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+(
+    gid        bigint unsigned auto_increment
+        primary key,
+    exchange   varchar(30)                                not null,
+    subAccount varchar(30)                                not null,
+    time       datetime(3)                                not null,
+    currency     varchar(12)                                not null,
+    balanceInUSD decimal(32, 8) unsigned default 0.00000000 not null,
+    balanceInBTC decimal(32, 8) unsigned default 0.00000000 not null,
```
Comment (c9s) [inline migrations/mysql/20211211034819_add_asset_balance_history_detail.sql:12]: use upper case for keywords? `NOT NULL` and `UNSIGNED DEFAULT`

File: pkg/exchange/ftx/exchange.go [hunk 11]
```diff
@@ -461,11 +488,68 @@ func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) erro
 }
 
 func (e *Exchange) QueryTicker(ctx context.Context, symbol string) (*types.Ticker, error) {
-	panic("implement me")
+	ticketMap, err := e.QueryTickers(ctx, symbol)
+
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:492]:
```
if ticker, ok := ticketMap[symbol] ; ok {
    return ticker
}
return nil, fmt.Errors("ticker %s not found", symbol)
```

File: pkg/exchange/ftx/exchange.go [hunk 12]
```diff
@@ -461,11 +488,68 @@ func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) erro
 }
 
 func (e *Exchange) QueryTicker(ctx context.Context, symbol string) (*types.Ticker, error) {
-	panic("implement me")
+	ticketMap, err := e.QueryTickers(ctx, symbol)
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:491]:
```
if err != nil {
    return nil, err
}
```

File: pkg/exchange/ftx/exchange.go [hunk 13]
```diff
@@ -461,11 +488,68 @@ func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) erro
...
+
+	markets, err := e._QueryMarkets(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string]struct{})
+	exists := struct{}{}
+	for _, s := range symbol {
+		m[toGlobalSymbol(s)] = exists
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:516]:
```
m[toGlobalSymbol(s)] = struct{}{}
```
you can use this

File: pkg/exchange/ftx/ticker_test.go [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestExchange_QueryTickers_AllSymbols(t *testing.T) {
+	key := os.Getenv("FTX_API_KEY")
+	secret := os.Getenv("FTX_API_SECRET")
+	subAccount := os.Getenv("FTX_SUBACCOUNT")
+	if len(key) == 0 && len(secret) == 0 {
+		t.Skip("api key/secret are not configured")
+		return
```
Comment (c9s) [inline pkg/exchange/ftx/ticker_test.go:17]: `return` can ignored, because t.Skip uses Fatal

File: pkg/exchange/ftx/exchange.go [hunk 15]
```diff
@@ -461,11 +488,68 @@ func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) erro
...
+
+		if err := requestLimit.Wait(ctx); err != nil {
+			logrus.WithError(err).Errorf("order rate limiter wait error")
+		}
+
+		//ctx context.Context, market string, interval types.Interval, limit int64, start, end time.Time
+		prices, err := rest.HistoricalPrices(ctx, v.Market.LocalSymbol, types.Interval1h, 1, time.Now().Add(time.Duration(-1)*time.Hour), time.Now())
+		if err != nil || !prices.Success || len(prices.Result) == 0 {
+			continue
+			//panic(fmt.Sprintf("ftx QueryTickers fail : %s ", k))
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:534]: remove this?

File: pkg/exchange/ftx/exchange.go [hunk 16]
```diff
@@ -461,11 +488,68 @@ func (e *Exchange) CancelOrders(ctx context.Context, orders ...types.Order) erro
...
+	m := make(map[string]struct{})
+	exists := struct{}{}
+	for _, s := range symbol {
+		m[toGlobalSymbol(s)] = exists
+	}
+
+	rest := e.newRest()
+
+	for k, v := range markets {
+		if _, ok := m[toGlobalSymbol(k)]; len(symbol) != 0 && !ok {
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:522]: 這邊是? 如果 map 沒有定義, 而且要查固定幾組 symbol 就跳過?

File: pkg/types/account.go [hunk 17]
```diff
@@ -41,25 +41,36 @@ func (b Balance) String() string {
...
-	InUSD      fixedpoint.Value `json:"inUSD"`
-	InBTC      fixedpoint.Value `json:"inBTC"`
-	Time       time.Time        `json:"time"`
+	Currency  string           `json:"currency"`
+	Total     fixedpoint.Value `json:"total"`
+	InUSD     fixedpoint.Value `json:"inUSD"`
+	InBTC     fixedpoint.Value `json:"inBTC"`
+	Time      time.Time        `json:"time"`
+	Lock      fixedpoint.Value `json:"lock"`
+	Available fixedpoint.Value `json:"available"`
```
Comment (c9s) [inline pkg/types/account.go:50]: Asset 需要紀錄 Lock 和 Avaiable 嗎?

Comment (c9s) [reply pkg/types/account.go:50]: 看起來直接用 Total 就好?

File: migrations/sqlite3/20211211034818_add_nav_history_details.sql [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    gid    bigint unsigned auto_increment PRIMARY KEY,
+    `exchange`          VARCHAR             NOT NULL DEFAULT '',
+    `subaccount`          VARCHAR             NOT NULL DEFAULT '',
+    time   DATETIME(3)     NOT NULL DEFAULT (strftime('%s','now')),
+    currency     VARCHAR                                NOT NULL,
+    balance_in_usd DECIMAL DEFAULT 0.00000000 NOT NULL,
+    balance_in_btc DECIMAL DEFAULT 0.00000000 NOT NULL,
+    balance    DECIMAL DEFAULT 0.00000000 NOT NULL,
+    available  DECIMAL DEFAULT 0.00000000 NOT NULL,
+    locked     DECIMAL DEFAULT 0.00000000 NOT NULL
```
Comment (c9s) [inline migrations/sqlite3/20211211034818_add_nav_history_details.sql:14]: available, locked 會需要紀錄嗎

File: pkg/types/account.go [hunk 19]
```diff
@@ -41,25 +41,36 @@ func (b Balance) String() string {
 }
 
 type Asset struct {
-	Currency   string           `json:"currency"`
-	Total      fixedpoint.Value `json:"total"`
-	InUSD      fixedpoint.Value `json:"inUSD"`
-	InBTC      fixedpoint.Value `json:"inBTC"`
-	Time       time.Time        `json:"time"`
+	Currency  string           `json:"currency"`
```
Comment (c9s) [inline pkg/types/account.go:44]: add db tag here, i mean `db:"currency"`, query 的時候比較容易 fill struct

File: pkg/service/account.go [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+func (s *AccountService) InsertAsset(time time.Time, name types.ExchangeName, account string, assets types.AssetMap) error {
+
+	if s.DB == nil {
+		//skip db insert when no db connection setting.
+		return nil
+	}
+
+	err := error(nil)
```
Comment (c9s) [inline pkg/service/account.go:24]: `var err error` 這樣寫就可以了

File: pkg/service/account.go [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+	for _, v := range assets {
+		_, _err := s.DB.Exec(`
+			insert into nav_history_details ( exchange, subaccount, time, currency, balance_in_usd, balance_in_btc,
+		                                              balance,available,locked)
+				values (?,?,?,?,?,?,?,?,?);
+		`, name, account, time, v.Currency, v.InUSD, v.InBTC, v.Total, v.Available, v.Lock)
+
+		if _err != nil {
+			err = _err
```
Comment (c9s) [inline pkg/service/account.go:33]: 可以用 multierr 紀錄多個 error

File: pkg/exchange/ftx/exchange.go [hunk 22]
```diff
@@ -35,6 +35,16 @@ type Exchange struct {
 	restEndpoint *url.URL
 }
 
+type Market struct {
```
Comment (c9s) [inline pkg/exchange/ftx/exchange.go:38]: How about MarketTicker?

**Comment by tony1223** [comment]
改用 step 2 那個 pr 處理

**Pull request closed without merging**